File: src/main/java/com/andrascsanyi/encyclopediagalactica/document/infra/mappers/DocumentMapper.java
package com.andrascsanyi.encyclopediagalactica.document.infra.mappers;

import com.andrascsanyi.encyclopediagalactica.document.contracts.DocumentInput;
import com.andrascsanyi.encyclopediagalactica.document.contracts.DocumentResult;
import com.andrascsanyi.encyclopediagalactica.document.entities.Document;
import org.mapstruct.InjectionStrategy;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.factory.Mappers;

@Mapper(componentModel = "spring", injectionStrategy = InjectionStrategy.CONSTRUCTOR)
public interface DocumentMapper {

    DocumentMapper INSTANCE = Mappers.getMapper(DocumentMapper.class);

    @Mapping(source = "id", target = "id")
    @Mapping(source = "name", target = "name")
    @Mapping(source = "description", target = "description")
    Document documentInputToDocument(DocumentInput documentInput);

    @Mapping(source = "id", target = "id")
    @Mapping(source = "name", target = "name")
    @Mapping(source = "description", target = "description")
    DocumentResult documentToDocumentInput(Document document);
}

File: src/dotnet/BusinessLogic/Mappers/DocumentMapper.cs
namespace EncyclopediaGalactica.BusinessLogic.Mappers;

using Contracts;
using Entities;

/// <inheritdoc />
public class DocumentMapper : IDocumentMapper
{
    /// <inheritdoc />
    public Document MapDocumentInputToDocument(DocumentInput input)
    {
        return new Document
        {
            Id = input.Id,
            Name = input.Name,
            Description = input.Description,
            Uri = input?.Uri
        };
    }

    /// <inheritdoc />
    public List<DocumentResult> MapDocumentsToDocumentResults(List<Document> l)
    {
        List<DocumentResult> resultList = new List<DocumentResult>();
        if (!l.Any()) return resultList;

        foreach (Document item in l)
        {
            resultList.Add(MapDocumentToDocumentResult(item));
        }

        return resultList;
    }

    /// <inheritdoc />
    public DocumentResult MapDocumentToDocumentResult(Document document)
    {
        return new DocumentResult
        {
            Id = document.Id,
            Name = document.Name,
            Description = document.Description,
            Uri = document?.Uri
        };
    }
}